Correct Error Message Level when Node Types are Undefined

RKStrand opened the pull request:
Pull request overview
---------------------

- Fixes #11061 

### Description of the purpose of this PR

In an example file in our test suite, there exists a water heater that does not have any nodes defined in its own object but it is part of a branch where nodes for this object are defined.  This conflict is noted with a warning message, but the simulation is allowed to continue with unknown consequences.  With this correction, the error message has been bumped up to a severe/fatal so that the user has to fix the problem/inconsistency in the input file.  In addition, the input file (ASHRAE901_OfficeSmall_STD2019_Denver.idf) has been corrected so that the branch and the water heater object both have the inlet and outlet node names.  A unit test that verifies the error occurrence and messages is also included.

In general, any connection error where the parent (in this case the Branch) has registered node connections that are not matched by a child object (in this case the water heater) leaving the node types unefined will now result in a severe/fatal error.

Previous error:
```
** Warning ** Potential Node Connection Error for object WATERHEATER:MIXED, name=SHWSYS1 WATER HEATER
** ~~~ ** Node Types are still UNDEFINED -- See Branch/Node Details file for further information
** ~~~ ** Inlet Node : SHWSYS1 PUMP-SHWSYS1 WATER HEATERNODE
** ~~~ ** Outlet Node: SHWSYS1 SUPPLY EQUIPMENT OUTLET NODE
```

Revised error:
```
   ** Severe  ** Potential Node Connection Error for object WATERHEATER:MIXED, name=SHWSYS1 WATER HEATER
   **   ~~~   **   Node Types are still UNDEFINED -- See Branch/Node Details file for further information
   **   ~~~   **   Inlet Node : SHWSYS1 PUMP-SHWSYS1 WATER HEATERNODE
   **   ~~~   **   Outlet Node: SHWSYS1 SUPPLY EQUIPMENT OUTLET NODE
   ************* There was 1 node connection error noted.
   **  Fatal  ** Please correct either the branch nodes or the component nodes so that they match.
```

### Pull Request Author

<!-- Add to this list or remove from it as applicable.  This is a simple templated set of guidelines. -->

 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [X] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer

<!-- This will not be exhaustively relevant to every PR. -->

 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [x] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

RKStrand committed:
Testing Out What Happens in the Test Suite

Just a little experiment with our test suite to see if there are any other failures of this kind...

RKStrand committed:
Correction of message and IDF

This commit corrects the temporary text of the fatal error.  It also corrects the input file that now causes a severe/fatal error.  This input file will have different results than develop because the node issue has been fixed in this input file.

RKStrand committed:
Unit Test for Error Message

This unit test causes the error message that was changed and tests to make sure that it is being correctly reported at the correct level.

RKStrand commented:
I should have added that the changes will result in differences in the input files.  The unrevised IDF will fatal out now rather than just have a warning message.  Also when comparing the old input file run with develop to the new input file run in the revised code, there will be differences in the output.  All other files should show no differences in the output.

mjwitte commented on src/EnergyPlus/SimulationManager.cc:
```diff
@@ -2256,12 +2256,13 @@ namespace SimulationManager {
                 ShowContinueError(state, "  Node Types are still UNDEFINED -- See Branch/Node Details file for further information");
                 ShowContinueError(state, format("  Inlet Node : {}", state.dataBranchNodeConnections->CompSets(Count).InletNodeName));
                 ShowContinueError(state, format("  Outlet Node: {}", state.dataBranchNodeConnections->CompSets(Count).OutletNodeName));
+                ShowFatalError(state, "Please correct either the branch nodes or the component nodes so that they match.");
```
@RKStrand When I test the original file (with missing nodes) the new code crashes. I'm not sure why/how, but following this in the debugger, `ShowFatalError` is being called multiple times. Adding a local error flag and moving the ShowFata to the end of `ReportLoopConnections` seems to fix this.

mjwitte replied:
Note that this change will require some changes to your unit test.

RKStrand replied:
Weird.  Not sure what happened there.  I'll implement the recommended change and update the unit test.  Look for the mods after the holiday weekend (hopefully).  Thanks!

RKStrand replied:
@mjwitte Latest commit fixes this problem.  Avoids a hard crash as per your suggestion.  Unit test also updated.

RKStrand committed:
Correction based on Reviewer Comment

Corrected a problem where the original IDF that revealed the problem was crashing after the fix was made.  This commit corrects the problem and updates the unit test to reflect this change.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 11061MissingNodesinWaterHeaterMixedErrorMessageLevel

mjwitte reviewed (approved):
Original defect file now produces this error:
```
   ** Severe  ** Potential Node Connection Error for object WATERHEATER:MIXED, name=SHWSYS1 WATER HEATER
   **   ~~~   **   Node Types are still UNDEFINED -- See Branch/Node Details file for further information
   **   ~~~   **   Inlet Node : SHWSYS1 PUMP-SHWSYS1 WATER HEATERNODE
   **   ~~~   **   Outlet Node: SHWSYS1 SUPPLY EQUIPMENT OUTLET NODE
   ************* There was 1 node connection error noted.
   **  Fatal  ** Please see severe error(s) and correct either the branch nodes or the component nodes so that they match.
```
I'll merge in develop and let CI run one more time, then merge.

Myoldmopar commented on src/EnergyPlus/SimulationManager.cc:
```diff
@@ -2723,6 +2725,10 @@ namespace SimulationManager {
+
+        if (nodeConnectionErrorFlag) {
+            ShowFatalError(state, "Please see severe error(s) and correct either the branch nodes or the component nodes so that they match.");
+        }
```
This is a fine fix, bumping it to a severe makes sense.  And adding the separate flag to provide extra error context is a good change.

Myoldmopar commented on testfiles/ASHRAE901_OfficeSmall_STD2019_Denver.idf:
```diff
@@ -5257,7 +5257,11 @@
+    BLDG_SWH_SCH,            !- Use Flow Rate Fraction Schedule Name
+    ,                        !- Cold Water Supply Temperature Schedule Name
+    SHWSys1 Pump-SHWSys1 Water HeaterNode, !- Use Side Inlet Node Name
+    SHWSys1 Supply Equipment Outlet Node;  !- Use Side Outlet Node Name
```
And yeah, good stuff getting the test file fixed up.

Myoldmopar commented on tst/EnergyPlus/unit/SimulationManager.unit.cc:
```diff
@@ -551,3 +552,34 @@ TEST_F(EnergyPlusFixture, Test_SimulationControl_PureLoadCalc)
+        "   ..... Last severe error=Potential Node Connection Error for object WATERHEATER:MIXED, name=WaterHeaterMixed1",
+    });
+
+    EXPECT_TRUE(compare_err_stream(error_string, true));
```
As a quick aside, there is a separate unit test function that you can call that is a bit easier.  In this case, we don't care what the *exact* error output looks like.  We just really want to make sure it is reporting the node connection error.  Instead of `compare_err_stream()`, try `compare_err_stream_substring()`.  In that function you could call:

```
EXPECT_TRUE(compare_err_stream_substring("Potential Node Connection Error", true));
```

and that should be sufficient.

RKStrand replied:
@Myoldmopar That is very interesting and not something that I realized was an option.  I will definitely try to remember that the next time I need to do something like this in a unit test.  Thanks!

Myoldmopar reviewed (approved):
Looks good, will check locally now.

Myoldmopar commented:
Passing happily here, merging.  Thanks @RKStrand
